server/region_syncer/history_buffer.go

```diff
@@ -0,0 +1,106 @@
+// copyright 2018 pingcap, inc.
+//
+// licensed under the apache license, version 2.0 (the "license");
+// you may not use this file except in compliance with the license.
+// you may obtain a copy of the license at
+//
+//     http://www.apache.org/licenses/license-2.0
+//
+// unless required by applicable law or agreed to in writing, software
+// distributed under the license is distributed on an "as is" basis,
+// see the license for the specific language governing permissions and
+// limitations under the license.
+
+package syncer
+
+import (
+	"strconv"
+
+	"github.com/pingcap/pd/server/core"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	historyKey        = "historyIndex"
+	defaultFlushCount = 100
+)
+
+type historyBuffer struct {
+	index      uint64
+	records    []*core.RegionInfo
+	head       int
+	tail       int
+	size       int
+	kv         core.KVBase
+	flushCount int
+}
+
+func newHistoryBuffer(size int, kv core.KVBase) *historyBuffer {
+	// use an empty space to simplify operation
+	size++
+	if size < 2 {
+		size = 2
+	}
+	records := make([]*core.RegionInfo, size)
+	h := &historyBuffer{
+		records: records,
+		size:    size,
+		kv:      kv,
+	}
+	h.reload()
+	return h
+}
+
+func (h *historyBuffer) len() int {
+	if h.tail < h.head {
+		return h.tail + h.size - h.head
```

Comment:
Is it always equal to `h.size - 1`?

Reply:
yes, if we do not use the operator like  `pop_front`.